tests/core/BearTest.py

```diff
@@ -0,0 +1,261 @@
+        uut.add_dependency_results(bear2, [4, 5, 6])
+
+        self.assertEqual(uut.dependency_results,
+                         {Bear1: [1, 2, 3], Bear2: [4, 5, 6]})
+
+        uut.add_dependency_results(bear1, [7, 8, 9])
+
+        self.assertEqual(uut.dependency_results,
+                         {Bear1: [1, 2, 3, 7, 8, 9], Bear2: [4, 5, 6]})
+
+    def test_execute_task(self):
+        # Test the default implementation of execute_task().
+        section = Section('test-section')
+        filedict = {}
+
+        uut = BearWithAnalysis(section, filedict)
+
+        results = uut.execute_task((10, 20), {'z': 30})
+
+        self.assertEqual(results, [10, 20, 30])
+
+    def test_check_prerequisites(self):
+        section = Section('test-section')
+        filedict = {}
+
+        # Test default implementation.
+        expected_possibilities = [
+            'Following requirements are not installed: '
+            '{} (installable via `{}`), {} (installable via `{}`)'.format(
+                req1.package, ' '.join(req1.install_command()),
+                req2.package, ' '.join(req2.install_command()))
+            for req1, req2 in permutations(BearWithPrerequisites.REQUIREMENTS)
+        ]
+
+        self.assertIn(BearWithPrerequisites.check_prerequisites(),
+                      expected_possibilities)
+
+        # Test correct exception throwing in constructor.
+        BearWithPrerequisitesOverride.prerequisites_fulfilled = True
+        BearWithPrerequisitesOverride(section, filedict)
+
+        BearWithPrerequisitesOverride.prerequisites_fulfilled = False
+        with self.assertRaises(RuntimeError) as cm:
+            BearWithPrerequisitesOverride(section, filedict)
+
+        self.assertEqual(
+            str(cm.exception),
+            'The bear BearWithPrerequisitesOverride does not fulfill all '
+            'requirements.')
+
+        BearWithPrerequisitesOverride.prerequisites_fulfilled = (
+            'This is on purpose due to running inside a test.')
+        with self.assertRaises(RuntimeError) as cm:
+            BearWithPrerequisitesOverride(section, filedict)
+
+        self.assertEqual(
+            str(cm.exception),
+            'The bear BearWithPrerequisitesOverride does not fulfill all '
+            'requirements. This is on purpose due to running inside a test.')
+
+    def test_get_metadata(self):
+        uut = BearWithAnalysis.get_metadata()
+        self.assertEqual(uut.non_optional_params,
+                         {'x': ('First value.', int),
+                          'y': ('Second value.', int)})
+        self.assertEqual(
+            uut.optional_params,
+            {'z': ("Third value. (Optional, defaults to '33'.)", int, 33)})
+
+    def test_get_non_optional_settings(self):
+        self.assertEqual(BearWithAnalysis.get_non_optional_settings(),
+                         {'x': ('First value.', int),
+                          'y': ('Second value.', int)})
+
```

Comment:
This function will never return then. It's simpler on purpose, because:
1. less code
2. less complicated
3. it's actually hard to create a circular dependency, you really have to create one purposely

See the old/current bear where we do this (nearly) the same